File: test_decision_page.py
# -*- coding: utf-8 -*-
"""처분 판정 페이지가 사람에게 무엇을 내미는가.

    python3 test_decision_page.py     # exit 0 = all scenarios pass

무엇이 답이 되는지는 `decision_page.js`가 정하고 `test_decision_page.mjs`가
봅니다. 여기서 붙잡는 것은 화면의 성질입니다: 이 페이지가 묻는 일만 카드가
되는가, 고르는 칸이 비어서 나가는가, 화면의 선택지와 논리의 어휘가 같은가.
"""
import csv
import io
import os
import re
import shutil
import sys
import tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)

import decision_page as D                                        # noqa: E402

N = [0]
FAIL = []


def check(name, ok, detail=""):
    N[0] += 1
    print("  %s %s%s" % ("ok  " if ok else "FAIL", name,
                         "" if ok else "  <- %s" % (detail,)))
    if not ok:
        FAIL.append(name)


def write(path, columns, rows):
    with io.open(path, "w", encoding="utf-8", newline="") as fh:
        w = csv.DictWriter(fh, fieldnames=columns)
        w.writeheader()
        for r in rows:
            w.writerow(dict((c, r.get(c, "")) for c in columns))


TMP = tempfile.mkdtemp(prefix="fdt-decision-")
RUN = os.path.join(TMP, "run")
PLANS = os.path.join(RUN, "plans")
os.makedirs(os.path.join(RUN, "crops"))
os.makedirs(PLANS)

UNC = "셀 수 없다고 하신 그림 — 처분 결정"
MIX = "캡션이 스스로 어긋난 그림 — 어느 패널이 데이터인지 확인"
GEOM = "패널 기하 (읽을 자리·눈금·표 종류)"

write(os.path.join(PLANS, D.FIGURES),
      ["Publication_ID", "Draft_ID", "Figure_Number", "Page", "Image",
       "Disposition", "Observed_Panel_Count", "Dispersion_Type",
       "Errorbar_Definition_Source", "Needs"],
      [{"Publication_ID": "A", "Draft_ID": "d1", "Figure_Number": "FIG1",
        "Page": "3", "Image": os.path.join("crops", "a1.png"),
        "Observed_Panel_Count": "", "Needs": UNC},
       {"Publication_ID": "A", "Draft_ID": "d2", "Figure_Number": "FIG2",
        "Page": "4", "Image": os.path.join("crops", "a1.png"),
        "Observed_Panel_Count": "4", "Dispersion_Type": "IQR",
        "Errorbar_Definition_Source": "Box plots indicate minimum,",
        "Needs": MIX},
       # 이 페이지가 묻지 않는 일. 단추로 답할 수 없습니다.
       {"Publication_ID": "B", "Draft_ID": "d3", "Figure_Number": "FIG1",
        "Page": "2", "Image": os.path.join("crops", "a1.png"),
        "Observed_Panel_Count": "6", "Needs": "%s · 오차 정의" % GEOM}])

write(os.path.join(RUN, D.COUNTS),
      ["Draft_ID", "Entry_Status", "Uncountable_Reason", "Objection_Reason"],
      [{"Draft_ID": "d1", "Entry_Status": "SEEN_UNCOUNTABLE",
        "Uncountable_Reason": "축이 겹쳐 어디까지가 한 패널인지 모르겠습니다"}])

write(os.path.join(RUN, D.CAPTIONS),
      ["Draft_ID", "Caption_Full"],
      [{"Draft_ID": "d2",
        "Caption_Full": "Fig. 2 | b Example of capillaroscopy. Box plots "
                        "indicate minimum, 25th percentile, median."}])

from PIL import Image                                            # noqa: E402
Image.new("RGB", (40, 30), (200, 40, 40)).save(
    os.path.join(RUN, "crops", "a1.png"))

HTML, ITEMS = D.build(RUN, PLANS, log=lambda *_a: None)

# REVERT: 계획서의 할 일을 전부 카드로 만든다. 패널 기하는 읽을 자리와 눈금을
# 쓰는 일이라 단추로 답할 수 없고, 오차 정의는 다른 페이지가 이미 묻습니다.
# 답할 수 없는 칸 앞에 사람을 앉히는 것은 묻지 않는 것보다 나쁩니다.
check("이 페이지가 묻는 일만 카드가 된다",
      ITEMS == 2 and "data-id='UNCOUNTABLE::d1'" in HTML
      and "data-id='MIXED::d2'" in HTML
      # base64 안에 아무 글자나 들어 있으므로 카드의 열쇠로만 봅니다.
      and "::d3'" not in HTML,
      ITEMS)

# REVERT: 화면의 선택지를 논리와 따로 적는다. 화면에만 있는 답은 고를 수는
# 있지만 답이 되지 않고, 논리에만 있는 답은 아무도 고를 수 없습니다.
_js = io.open(os.path.join(HERE, D.LOGIC), encoding="utf-8").read()
_ok = True
for kind, pairs in D.LABELS.items():
    m = re.search(r"%s:\s*\[([^\]]*)\]" % kind, _js)
    want = [v.strip().strip("'") for v in m.group(1).split(",")] if m else []
    if [v for v, _l in pairs] != want:
        _ok = False
check("화면의 선택지와 논리의 어휘가 같다", _ok,
      dict((k, [v for v, _l in p]) for k, p in D.LABELS.items()))

check("물음마다 받는 답이 다르다",
      "value='COUNTABLE'" in HTML and "value='PARTIAL'" in HTML
      and HTML.count("value='COUNTABLE'") == 1)

# REVERT: 고른 채로 낸다. 사람이 하는 일이 판정이 아니라 동의가 됩니다.
_inputs = [l for l in HTML.splitlines() if "<input type=" in l]
check("아무 선택지도 골라 둔 채로 나가지 않는다",
      bool(_inputs) and all("checked" not in l for l in _inputs),
      [l for l in _inputs if "checked" in l][:1])

# REVERT: 계수 시트에 적어 두신 말을 싣지 않는다.
check("사람이 적어 둔 말이 카드에 실린다",
      "축이 겹쳐 어디까지가 한 패널인지" in HTML)

# REVERT: 어긋난 캡션에 캡션을 싣지 않는다. 무엇이 어긋났는지 보지 못한 채
# 어느 패널이 데이터인지 정하게 됩니다.
check("어긋난 캡션 카드에는 캡션 전문이 실린다",
      "Example of capillaroscopy" in HTML and "Box plots indicate" in HTML)

check("META가 물음과 나갈 이름을 들고 있다",
      '"MIXED::d2": {"kind": "MIXED", "draft": "d2"}' in HTML,
      [l for l in HTML.splitlines() if l.startswith("var META")][:1])

check("결정은 페이지 안에 실려 있다",
      "function decisionOf" in HTML and "function buildCsv" in HTML)

_empty = os.path.join(TMP, "empty")
os.makedirs(_empty)
write(os.path.join(_empty, D.FIGURES), ["Draft_ID", "Needs"],
      [{"Draft_ID": "d1", "Needs": GEOM}])
try:
    D.build(RUN, _empty, log=lambda *_a: None)
    _stopped = False
except SystemExit:
    _stopped = True
check("물을 것이 없으면 빈 페이지를 내지 않고 멈춘다", _stopped)

shutil.rmtree(TMP, ignore_errors=True)
print()
print("FDT_SCENARIOS_RUN=%d" % N[0])
print("%d scenarios run" % N[0])
if FAIL:
    print("%d FAILED: %s" % (len(FAIL), FAIL))
    raise SystemExit(1)
print("all scenarios passed")

File: test_decision_page.mjs
/* 처분 판정 페이지의 결정들에 대한 시나리오.
 *
 *     node test_decision_page.mjs
 */
import assert from 'node:assert/strict';
import { createRequire } from 'node:module';
const L = createRequire(import.meta.url)('./decision_page.js');

let pass = 0; const fails = [];
function test(name, fn) {
  try { fn(); console.log('ok    ' + name); pass++; }
  catch (e) { console.log('FAIL  ' + name + '  <- ' + e.message); fails.push(name); }
}

function state(over) {
  return Object.assign({ kind: 'UNCOUNTABLE', choice: 'RECROP', which: '',
                         note: '', seen: true, draft: 'd1' }, over || {});
}

test('다 갖춘 답은 한 줄이 된다', () => {
  const got = L.decisionOf('d1', state());
  assert.equal(got.ready, true);
  assert.equal(got.row.Decision, 'RECROP');
  assert.equal(got.row.Seen_By_Person, '1');
});

/* REVERT: 그림을 보았다고 누르지 않아도 답으로 친다. 고르기만 한 줄이 "내가
 * 이 그림을 보았다"는 진술로 나갑니다 - 이 페이지가 있는 이유가 사라집니다. */
test('그림을 봤다고 누르지 않으면 답이 아니다', () => {
  const got = L.decisionOf('d1', state({ seen: false }));
  assert.equal(got.ready, false);
  assert.match(got.why, /직접 보셨다/);
  assert.equal(got.row, null);
});

/* REVERT: 물음마다 받을 답을 가르지 않는다. "셀 수 없다"에 DATA가, 어긋난
 * 캡션에 COUNTABLE이 답으로 들어가고, 그것을 읽는 쪽은 무엇을 물었는지
 * 모르는 채 처분만 받습니다. */
test('물음에 없는 답은 이름을 대며 거절한다', () => {
  const got = L.decisionOf('d1', state({ kind: 'UNCOUNTABLE', choice: 'DATA' }));
  assert.equal(got.ready, false);
  assert.match(got.why, /DATA/);
});
test('물음마다 받는 답이 다르다', () => {
  assert.equal(L.decisionOf('d1', state({ kind: 'MIXED', choice: 'DATA' })).ready, true);
  assert.equal(L.decisionOf('d1', state({ kind: 'CROP_DISPUTE', choice: 'COUNTABLE' })).ready, false);
});
test('무엇을 묻는지 모르는 줄은 답이 되지 않는다', () => {
  const got = L.decisionOf('d1', state({ kind: '' }));
  assert.equal(got.ready, false);
  assert.match(got.why, /모르는/);
});
/* REVERT: 빈칸을 "쓸 수 없는 답"으로 흘려보낸다. 어휘 검사가 어차피 막지만,
 * 아직 고르지 않은 사람에게 "이 물음에 쓸 수 없는 답입니다: "라고 말하게
 * 됩니다 - 자기가 무엇을 잘못했는지 알 수 없는 말입니다. 이 가드가 지키는
 * 것은 답의 모양이 아니라 사람에게 하는 말입니다. */
test('고르지 않으면 아직 고르지 않았다고 말한다', () => {
  const got = L.decisionOf('d1', state({ choice: '' }));
  assert.equal(got.ready, false);
  assert.match(got.why, /아직 고르지 않았습니다/);
});

/* REVERT: 어느 패널인지 묻지 않는다. "일부 패널만 데이터"라는 답에서 어느
 * 패널인지 빠지면, 다음 사람은 그 그림 앞에서 처음부터 다시 봅니다. */
test('일부 패널만이라면 어느 패널인지 적어야 한다', () => {
  const got = L.decisionOf('d1', state({ kind: 'MIXED', choice: 'PARTIAL' }));
  assert.equal(got.ready, false);
  assert.match(got.why, /어느 패널/);
  const ok = L.decisionOf('d1', state({ kind: 'MIXED', choice: 'PARTIAL', which: 'a, c' }));
  assert.equal(ok.ready, true);
  assert.equal(ok.row.Which_Panels, 'a, c');
});
test('패널을 묻지 않는 답에는 패널 칸을 싣지 않는다', () => {
  const got = L.decisionOf('d1', state({ kind: 'MIXED', choice: 'DATA', which: 'a' }));
  assert.equal(got.row.Which_Panels, '');
});

test('아직 못 정하겠다도 답이다', () => {
  const got = L.decisionOf('d1', state({ choice: 'HOLD' }));
  assert.equal(got.ready, true);
  assert.equal(got.row.Decision, 'HOLD');
});
test('그래도 봤다는 표시는 HOLD에도 필요하다', () => {
  assert.equal(L.decisionOf('d1', state({ choice: 'HOLD', seen: false })).ready, false);
});

/* REVERT: 열쇠를 그대로 `Draft_ID`로 내보낸다. 화면이 상태를 담는 열쇠와
 * 파이프라인이 아는 이름이 같으리라는 보장이 없습니다. */
test('나가는 이름은 상태가 들고 있는 Draft_ID다', () => {
  assert.equal(L.decisionOf('KEY', state({ draft: 'd9' })).row.Draft_ID, 'd9');
});

test('CSV는 답이 된 줄만 싣는다', () => {
  const csv = L.buildCsv(['a', 'b'],
                         { a: state({ draft: 'a' }),
                           b: state({ draft: 'b', seen: false }) }).split('\n');
  assert.equal(csv.length, 2);
  assert.match(csv[1], /^"a"/);
});
/* REVERT: `Object.keys(states)`를 목록으로 쓴다. 화면이 이름을 한 번 잘못
 * 읽으면 그 이름이 그대로 `Draft_ID`가 되어 나갑니다. */
test('이 페이지가 모르는 이름은 줄이 되지 않는다', () => {
  const csv = L.buildCsv(['a'], { a: state({ draft: 'a' }),
                                  '어디서 온 이름': state({ draft: 'x' }) }).split('\n');
  assert.equal(csv.length, 2);
});
test('따옴표가 든 메모도 그대로 나간다', () => {
  const csv = L.buildCsv(['a'], { a: state({ draft: 'a', note: '패널 "b"만' }) });
  assert.ok(csv.indexOf('""b""') >= 0, csv);
});

test('남은 일은 답이 안 된 그림을 센다', () => {
  assert.equal(L.remaining(['a', 'b', 'c'],
                           { a: state(), b: state({ seen: false }) }), 2);
});
test('보류는 남은 일이 아니라 따로 센다', () => {
  const st = { a: state({ choice: 'HOLD' }), b: state() };
  assert.equal(L.remaining(['a', 'b'], st), 0);
  assert.equal(L.held(['a', 'b'], st), 1);
});

console.log('');
console.log(pass + '/' + (pass + fails.length) + ' passed');
if (fails.length) { console.log('FAILED: ' + fails.join(', ')); process.exit(1); }
